// @antv/gpt-vis@0.4.7/dist/cjs/Text/config.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/Text/config.tsx
var config_exports = {};
__export(config_exports, {
  DEFAULT_VIS_TEXT_CONFIG: () => DEFAULT_VIS_TEXT_CONFIG,
  STATICS_KEY: () => STATICS_KEY
});
module.exports = __toCommonJS(config_exports);
var import_custom_icons = require("./custom-icons");
var import_mini_charts = require("./mini-charts");
var import_theme = require("./theme");
var STATICS_KEY = "__statics";
var DEFAULT_VIS_TEXT_CONFIG = {
  [STATICS_KEY]: {
    "icon:arrow-up": import_custom_icons.ArrowUp,
    "icon:arrow-down": import_custom_icons.ArrowDown,
    "mini-chart:proportion": import_mini_charts.ProportionChart,
    // TODO @羽熙 当前仅支持单折线图，之后可以考虑支持多折线图，在组件内实现分支判断
    "mini-chart:line": import_mini_charts.SingleLineChart
  },
  metric_name: {
    color: import_theme.TEXT_THEME.light.default88Color,
    fontWeight: 500
  },
  metric_value: {
    color: import_theme.TEXT_THEME.light.primaryColor
  },
  other_metric_value: {
    color: import_theme.TEXT_THEME.light.default65Color
  },
  delta_value: {
    color: import_theme.TEXT_THEME.light.default65Color
  },
  ratio_value: {
    color: import_theme.TEXT_THEME.light.default65Color
  },
  delta_value_pos: {
    color: import_theme.TEXT_THEME.light.posColor,
    prefix: "+"
  },
  delta_value_neg: {
    color: import_theme.TEXT_THEME.light.negColor,
    prefix: "-"
  },
  ratio_value_pos: {
    color: import_theme.TEXT_THEME.light.posColor,
    prefix: "icon:arrow-up"
  },
  ratio_value_neg: {
    color: import_theme.TEXT_THEME.light.negColor,
    prefix: "icon:arrow-down"
  },
  contribute_ratio: {
    color: import_theme.TEXT_THEME.light.conclusionColor
  },
  trend_desc: {
    color: import_theme.TEXT_THEME.light.conclusionColor,
    suffix: "mini-chart:line"
  },
  dim_value: {
    color: import_theme.TEXT_THEME.light.default88Color
  },
  time_desc: {
    color: import_theme.TEXT_THEME.light.default88Color
  },
  proportion: {
    suffix: "mini-chart:proportion"
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  DEFAULT_VIS_TEXT_CONFIG,
  STATICS_KEY
});
