// @antv/gpt-vis@0.4.7/dist/cjs/Text/theme.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/Text/theme.ts
var theme_exports = {};
__export(theme_exports, {
  TEXT_THEME: () => TEXT_THEME
});
module.exports = __toCommonJS(theme_exports);
var TEXT_THEME = {
  fontSizeBase: 14,
  chart: {
    proportionShadowColor: "#CDDDFD",
    proportionFillColor: "#3471F9",
    lineStrokeColor: "#5B8FF9"
  },
  light: {
    primaryColor: "#1677FF",
    default88Color: "rgba(0, 0, 0, 0.88)",
    default65Color: "rgba(0, 0, 0, 0.65)",
    posColor: "#FA541C",
    negColor: "#13A8A8",
    conclusionColor: "#1F0352"
  },
  dark: {
    primaryColor: "#4B91FF",
    default88Color: "rgba(255, 255, 255, 0.88)",
    default65Color: "rgba(255, 255, 255, 0.65)",
    posColor: "#FA541C",
    negColor: "#13A8A8",
    conclusionColor: "#D8C3F3"
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  TEXT_THEME
});
