// @antv/larkmap@1.5.1/es/components/Control/ZoomControl/index.js
function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }
function _iterableToArrayLimit(r, l) { var t = null == r ? null : "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (null != t) { var e, n, i, u, a = [], f = !0, o = !1; try { if (i = (t = t.call(r)).next, 0 === l) { if (Object(t) !== t) return; f = !1; } else for (; !(f = (e = i.call(t)).done) && (a.push(e.value), a.length !== l); f = !0); } catch (r) { o = !0, n = r; } finally { try { if (!f && null != t.return && (u = t.return(), Object(u) !== u)) return; } finally { if (o) throw n; } } return a; } }
function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }
import { Zoom as L7Zoom } from '@antv/l7';
import { omitBy } from 'lodash-es';
import React, { useEffect, useMemo, useState } from 'react';
import { getStyleText } from "../../../utils";
import { useScene } from "../../LarkMap/hooks";
import { useL7ComponentEvent, useL7ComponentPortal, useL7ComponentUpdate } from "../hooks";
export var ZoomControl = function ZoomControl(_ref) {
  var onShow = _ref.onShow,
    onHide = _ref.onHide,
    onAdd = _ref.onAdd,
    onRemove = _ref.onRemove,
    position = _ref.position,
    name = _ref.name,
    className = _ref.className,
    style = _ref.style,
    zoomInText = _ref.zoomInText,
    zoomOutText = _ref.zoomOutText,
    zoomInTitle = _ref.zoomInTitle,
    zoomOutTitle = _ref.zoomOutTitle,
    showZoom = _ref.showZoom;
  var scene = useScene();
  var _useState = useState(),
    _useState2 = _slicedToArray(_useState, 2),
    control = _useState2[0],
    setControl = _useState2[1];
  var styleText = useMemo(function () {
    return getStyleText(style);
  }, [style]);
  var _useL7ComponentPortal = useL7ComponentPortal(zoomInText),
    zoomInTextPortal = _useL7ComponentPortal.portal,
    zoomInTextDOM = _useL7ComponentPortal.dom;
  var _useL7ComponentPortal2 = useL7ComponentPortal(zoomOutText),
    zoomOutTextPortal = _useL7ComponentPortal2.portal,
    zoomOutTextDOM = _useL7ComponentPortal2.dom;
  var controlOptions = useMemo(function () {
    return {
      position: position,
      name: name,
      className: className,
      zoomInTitle: zoomInTitle,
      zoomOutTitle: zoomOutTitle,
      showZoom: showZoom,
      style: styleText,
      zoomInText: zoomInTextDOM,
      zoomOutText: zoomOutTextDOM
    };
  }, [position, name, className, styleText, zoomInTitle, zoomOutTitle, zoomInTextDOM, zoomOutTextDOM, showZoom]);
  useEffect(function () {
    var zoom = new L7Zoom(omitBy(controlOptions, function (value) {
      return value === undefined;
    }));
    setControl(zoom);
    scene.addControl(zoom);
    return function () {
      scene.removeControl(zoom);
      setControl(undefined);
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
  useL7ComponentUpdate(control, controlOptions);
  useL7ComponentEvent(control, {
    add: onAdd,
    remove: onRemove,
    show: onShow,
    hide: onHide
  });
  return /*#__PURE__*/React.createElement(React.Fragment, null, zoomInTextPortal, zoomOutTextPortal);
};
