// @antv/larkmap@1.5.1/lib/components/Control/MapThemeControl/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MapThemeControl = void 0;
var _l = require("@antv/l7");
var _lodash = require("lodash");
var _react = _interopRequireWildcard(require("react"));
var _utils = require("../../../utils");
var _hooks = require("../../LarkMap/hooks");
var _hooks2 = require("../hooks");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
const MapThemeControl = ({
  popperPlacement,
  btnText,
  title,
  vertical,
  position,
  className,
  options,
  popperTrigger,
  popperClassName,
  btnIcon,
  style,
  onAdd,
  onRemove,
  onShow,
  onHide,
  onPopperShow,
  onPopperHide,
  onSelectChange
}) => {
  const scene = (0, _hooks.useScene)();
  const [control, setControl] = (0, _react.useState)();
  const styleText = (0, _react.useMemo)(() => (0, _utils.getStyleText)(style), [style]);
  const {
    portal: btnIconPortal,
    dom: btnIconDOM
  } = (0, _hooks2.useL7ComponentPortal)(btnIcon);
  const controlOptions = (0, _react.useMemo)(() => {
    return {
      popperPlacement,
      popperTrigger,
      popperClassName,
      btnText,
      title,
      vertical,
      position,
      className,
      options,
      style: styleText,
      btnIcon: btnIconDOM
    };
  }, [options, popperPlacement, popperTrigger, popperClassName, btnText, title, vertical, position, className, styleText, btnIconDOM]);
  (0, _react.useEffect)(() => {
    const mapTheme = new _l.MapTheme((0, _lodash.omitBy)(controlOptions, value => value === undefined));
    setControl(mapTheme);
    scene.addControl(mapTheme);
    return () => {
      scene.removeControl(mapTheme);
      setControl(undefined);
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
  (0, _hooks2.useL7ComponentUpdate)(control, controlOptions);
  (0, _hooks2.useL7ComponentEvent)(control, {
    add: onAdd,
    remove: onRemove,
    show: onShow,
    hide: onHide,
    popperShow: onPopperShow,
    popperHide: onPopperHide,
    selectChange: onSelectChange
  });
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, btnIconPortal);
};
exports.MapThemeControl = MapThemeControl;
