// @antv/larkmap@1.5.1/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  LarkMap: true,
  Template: true,
  Marker: true,
  Popup: true,
  LayerPopup: true,
  ContextMenu: true,
  LocationSearch: true,
  RegionLocation: true,
  LegendIcon: true,
  LegendProportion: true,
  LegendCategories: true,
  LegendRamp: true,
  useDraw: true,
  useDrawGroup: true,
  version: true,
  syncScene: true,
  SyncScene: true
};
Object.defineProperty(exports, "ContextMenu", {
  enumerable: true,
  get: function () {
    return _ContextMenu.ContextMenu;
  }
});
Object.defineProperty(exports, "LarkMap", {
  enumerable: true,
  get: function () {
    return _LarkMap.LarkMap;
  }
});
Object.defineProperty(exports, "LayerPopup", {
  enumerable: true,
  get: function () {
    return _LayerPopup.LayerPopup;
  }
});
Object.defineProperty(exports, "LegendCategories", {
  enumerable: true,
  get: function () {
    return _LegendCategories.LegendCategories;
  }
});
Object.defineProperty(exports, "LegendIcon", {
  enumerable: true,
  get: function () {
    return _LegendIcon.LegendIcon;
  }
});
Object.defineProperty(exports, "LegendProportion", {
  enumerable: true,
  get: function () {
    return _LegendProportion.LegendProportion;
  }
});
Object.defineProperty(exports, "LegendRamp", {
  enumerable: true,
  get: function () {
    return _LegendRamp.LegendRamp;
  }
});
Object.defineProperty(exports, "LocationSearch", {
  enumerable: true,
  get: function () {
    return _LocationSearch.LocationSearch;
  }
});
Object.defineProperty(exports, "Marker", {
  enumerable: true,
  get: function () {
    return _Marker.Marker;
  }
});
Object.defineProperty(exports, "Popup", {
  enumerable: true,
  get: function () {
    return _Popup.Popup;
  }
});
Object.defineProperty(exports, "RegionLocation", {
  enumerable: true,
  get: function () {
    return _RegionLocation.RegionLocation;
  }
});
Object.defineProperty(exports, "SyncScene", {
  enumerable: true,
  get: function () {
    return _SyncScene.SyncScene;
  }
});
Object.defineProperty(exports, "Template", {
  enumerable: true,
  get: function () {
    return _Template.Template;
  }
});
Object.defineProperty(exports, "syncScene", {
  enumerable: true,
  get: function () {
    return _SyncScene.syncScene;
  }
});
Object.defineProperty(exports, "useDraw", {
  enumerable: true,
  get: function () {
    return _useDraw.useDraw;
  }
});
Object.defineProperty(exports, "useDrawGroup", {
  enumerable: true,
  get: function () {
    return _useDrawGroup.useDrawGroup;
  }
});
Object.defineProperty(exports, "version", {
  enumerable: true,
  get: function () {
    return _version.default;
  }
});
var _LarkMap = require("./components/LarkMap");
var _hooks = require("./components/LarkMap/hooks");
Object.keys(_hooks).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _hooks[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _hooks[key];
    }
  });
});
var _Layers = require("./components/Layers");
Object.keys(_Layers).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _Layers[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _Layers[key];
    }
  });
});
var _Control = require("./components/Control");
Object.keys(_Control).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _Control[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _Control[key];
    }
  });
});
var _Template = require("./components/Template");
var _Marker = require("./components/Marker");
var _Popup = require("./components/Popup");
var _LayerPopup = require("./components/LayerPopup");
var _ContextMenu = require("./components/ContextMenu");
var _LocationSearch = require("./components/LocationSearch");
var _RegionLocation = require("./components/RegionLocation");
var _LegendIcon = require("./components/Legend/LegendIcon");
var _LegendProportion = require("./components/Legend/LegendProportion");
var _LegendCategories = require("./components/Legend/LegendCategories");
var _LegendRamp = require("./components/Legend/LegendRamp");
var _useDraw = require("./components/Draw/use-draw");
var _useDrawGroup = require("./components/Draw/use-draw-group");
var _version = _interopRequireDefault(require("./version"));
var _SyncScene = require("./components/SyncScene");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
