// @antv/l7-composite-layers@0.17.1/dist/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FlowLayer = exports.IconImageLayer = exports.IconFontLayer = exports.ChoroplethLayer = exports.BubbleLayer = exports.PolygonLayer = exports.ImageLayer = exports.RasterLayer = exports.LineLayer = exports.HeatmapLayer = exports.TextLayer = exports.PointLayer = exports.LayerGroup = exports.version = void 0;
const tslib_1 = require("tslib");
// organize-imports-ignore
var version_1 = require("./version");
Object.defineProperty(exports, "version", { enumerable: true, get: function () { return tslib_1.__importDefault(version_1).default; } });
/** 类型定义导出 **/
tslib_1.__exportStar(require("./types"), exports);
/** 功能图层 **/
// 图层组及类型定义 | author by [yunji]](https://github.com/lvisei)
var layer_group_1 = require("./core/layer-group");
Object.defineProperty(exports, "LayerGroup", { enumerable: true, get: function () { return layer_group_1.LayerGroup; } });
/** 核心图层 **/
// 点图层及类型定义 | author by [yunji]](https://github.com/lvisei)
var point_layer_1 = require("./core-layers/point-layer");
Object.defineProperty(exports, "PointLayer", { enumerable: true, get: function () { return point_layer_1.PointLayer; } });
tslib_1.__exportStar(require("./core-layers/point-layer/types"), exports);
// 文本图层及类型定义 | author by [yunji]](https://github.com/lvisei)
var text_layer_1 = require("./core-layers/text-layer");
Object.defineProperty(exports, "TextLayer", { enumerable: true, get: function () { return text_layer_1.TextLayer; } });
tslib_1.__exportStar(require("./core-layers/text-layer/types"), exports);
// 热力图层及类型定义 | author by [yunji]](https://github.com/lvisei)
var heatmap_layer_1 = require("./core-layers/heatmap-layer");
Object.defineProperty(exports, "HeatmapLayer", { enumerable: true, get: function () { return heatmap_layer_1.HeatmapLayer; } });
tslib_1.__exportStar(require("./core-layers/heatmap-layer/types"), exports);
// 线图层及类型定义 | author by [yunji]](https://github.com/lvisei)
var line_layer_1 = require("./core-layers/line-layer");
Object.defineProperty(exports, "LineLayer", { enumerable: true, get: function () { return line_layer_1.LineLayer; } });
tslib_1.__exportStar(require("./core-layers/line-layer/types"), exports);
// 栅格图层类型及定义 ｜ author by [shihui](https://github.com/yiiiiiiqianyao)
var raster_layer_1 = require("./core-layers/raster-layer");
Object.defineProperty(exports, "RasterLayer", { enumerable: true, get: function () { return raster_layer_1.RasterLayer; } });
tslib_1.__exportStar(require("./core-layers/raster-layer/types"), exports);
// 图片图层类型及定义 ｜ author by [shihui](https://github.com/yiiiiiiqianyao)
var image_layer_1 = require("./core-layers/image-layer");
Object.defineProperty(exports, "ImageLayer", { enumerable: true, get: function () { return image_layer_1.ImageLayer; } });
tslib_1.__exportStar(require("./core-layers/image-layer/types"), exports);
// 面图层及类型定义 | author by [yunji]](https://github.com/lvisei)
var polygon_layer_1 = require("./core-layers/polygon-layer");
Object.defineProperty(exports, "PolygonLayer", { enumerable: true, get: function () { return polygon_layer_1.PolygonLayer; } });
tslib_1.__exportStar(require("./core-layers/polygon-layer/types"), exports);
/** 复合图层 **/
// 气泡图层及类型定义 | author by [yunji]](https://github.com/lvisei)
var bubble_layer_1 = require("./composite-layers/bubble-layer");
Object.defineProperty(exports, "BubbleLayer", { enumerable: true, get: function () { return bubble_layer_1.BubbleLayer; } });
tslib_1.__exportStar(require("./composite-layers/bubble-layer/types"), exports);
// 区域图层及类型定义 | author by [yunji]](https://github.com/lvisei)
var choropleth_layer_1 = require("./composite-layers/choropleth-layer");
Object.defineProperty(exports, "ChoroplethLayer", { enumerable: true, get: function () { return choropleth_layer_1.ChoroplethLayer; } });
tslib_1.__exportStar(require("./composite-layers/choropleth-layer/types"), exports);
// 标注图层及类型定义 | author by [lzxue]](https://github.com/lzxue)
var icon_layer_1 = require("./composite-layers/icon-layer");
Object.defineProperty(exports, "IconFontLayer", { enumerable: true, get: function () { return icon_layer_1.IconFontLayer; } });
Object.defineProperty(exports, "IconImageLayer", { enumerable: true, get: function () { return icon_layer_1.IconImageLayer; } });
tslib_1.__exportStar(require("./composite-layers/icon-layer/types"), exports);
// 客流聚合图层及类型定义 | author by [heiyexing]](https://github.com/heiyexing)
var flow_layer_1 = require("./composite-layers/flow-layer");
Object.defineProperty(exports, "FlowLayer", { enumerable: true, get: function () { return flow_layer_1.FlowLayer; } });
