// @antv/gpt-vis@0.4.7/dist/esm/Radar/index.js
import { Radar as ADCRadar } from '@ant-design/plots';
import { get } from 'lodash';
import React from 'react';
import { usePlotConfig } from "../ConfigProvider/hooks";
var defaultConfig = function defaultConfig(props) {
  var data = props.data,
    _props$xField = props.xField,
    xField = _props$xField === void 0 ? 'name' : _props$xField,
    _props$yField = props.yField,
    yField = _props$yField === void 0 ? 'value' : _props$yField;
  var hasGroupField = get(data, '[0].group') !== undefined;
  return {
    xField: xField,
    yField: yField,
    colorField: hasGroupField ? 'group' : undefined,
    area: {
      style: {
        fillOpacity: 0.5
      }
    },
    scale: {
      x: {
        padding: 0.6,
        align: 0
      },
      y: {
        nice: true
      }
    },
    axis: {
      x: {
        title: false,
        grid: true
      },
      y: {
        zIndex: 1,
        title: false,
        gridConnect: 'line',
        gridLineWidth: 1
      }
    },
    tooltip: function tooltip(d) {
      return {
        name: d[xField],
        value: d[yField]
      };
    }
  };
};
var Radar = function Radar(props) {
  var config = usePlotConfig('Radar', defaultConfig, props);
  return /*#__PURE__*/React.createElement(ADCRadar, config);
};
export default Radar;
