// @antv/larkmap@1.5.1/lib/components/Popup/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Popup = void 0;
var _l = require("@antv/l7");
var _lodash = require("lodash");
var _react = _interopRequireWildcard(require("react"));
var _utils = require("../../utils");
var _hooks = require("../Control/hooks");
var _hooks2 = require("../LarkMap/hooks");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
const Popup = ({
  style,
  closeButton,
  closeButtonOffsets,
  closeOnClick,
  closeOnEsc,
  maxWidth,
  anchor,
  offsets,
  stopPropagation,
  autoPan,
  autoClose,
  followCursor,
  className,
  lngLat,
  children,
  title,
  onOpen,
  onClose,
  onShow,
  onHide
}) => {
  const scene = (0, _hooks2.useScene)();
  const [popup, setPopup] = (0, _react.useState)();
  const styleText = (0, _react.useMemo)(() => (0, _utils.getStyleText)(style), [style]);
  const {
    portal: childrenPartial,
    dom: childrenDOM
  } = (0, _hooks.useL7ComponentPortal)(children);
  const {
    portal: titlePartial,
    dom: titleDOM
  } = (0, _hooks.useL7ComponentPortal)(title);
  const popupOptions = (0, _react.useMemo)(() => ({
    style: styleText,
    closeButton,
    closeButtonOffsets,
    closeOnClick,
    closeOnEsc,
    maxWidth,
    anchor,
    offsets,
    stopPropagation,
    autoPan,
    autoClose,
    followCursor,
    className,
    lngLat,
    html: childrenDOM,
    title: titleDOM
  }),
  // eslint-disable-next-line react-hooks/exhaustive-deps
  [styleText, closeButton,
  // eslint-disable-next-line react-hooks/exhaustive-deps
  JSON.stringify(closeButtonOffsets), closeOnClick, closeOnEsc, maxWidth, anchor,
  // eslint-disable-next-line react-hooks/exhaustive-deps
  JSON.stringify(offsets), stopPropagation, autoPan, autoClose, followCursor, className, childrenDOM, titleDOM,
  // eslint-disable-next-line react-hooks/exhaustive-deps
  JSON.stringify(lngLat)]);
  (0, _react.useEffect)(() => {
    const newPopup = new _l.Popup((0, _lodash.omitBy)(popupOptions, value => value === undefined));
    setPopup(newPopup);
    scene.addPopup(newPopup);
    return () => {
      scene.removePopup(newPopup);
      setPopup(undefined);
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
  (0, _hooks.useL7ComponentUpdate)(popup, popupOptions);
  (0, _hooks.useL7ComponentEvent)(popup, {
    open: onOpen,
    close: onClose,
    show: onShow,
    hide: onHide
  });
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, childrenPartial, titlePartial);
};
exports.Popup = Popup;
