// @antv/l7-draw@3.1.5/es/render/scene-render.js
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
function _classCallCheck(a, n) { if (!(a instanceof n)) throw new TypeError("Cannot call a class as a function"); }
function _defineProperties(e, r) { for (var t = 0; t < r.length; t++) { var o = r[t]; o.enumerable = o.enumerable || !1, o.configurable = !0, "value" in o && (o.writable = !0), Object.defineProperty(e, _toPropertyKey(o.key), o); } }
function _createClass(e, r, t) { return r && _defineProperties(e.prototype, r), t && _defineProperties(e, t), Object.defineProperty(e, "prototype", { writable: !1 }), e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
function _callSuper(t, o, e) { return o = _getPrototypeOf(o), _possibleConstructorReturn(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], _getPrototypeOf(t).constructor) : o.apply(t, e)); }
function _possibleConstructorReturn(t, e) { if (e && ("object" == _typeof(e) || "function" == typeof e)) return e; if (void 0 !== e) throw new TypeError("Derived constructors may only return object or undefined"); return _assertThisInitialized(t); }
function _assertThisInitialized(e) { if (void 0 === e) throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); return e; }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
function _getPrototypeOf(t) { return _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf.bind() : function (t) { return t.__proto__ || Object.getPrototypeOf(t); }, _getPrototypeOf(t); }
function _inherits(t, e) { if ("function" != typeof e && null !== e) throw new TypeError("Super expression must either be null or a function"); t.prototype = Object.create(e && e.prototype, { constructor: { value: t, writable: !0, configurable: !0 } }), Object.defineProperty(t, "prototype", { writable: !1 }), e && _setPrototypeOf(t, e); }
function _setPrototypeOf(t, e) { return _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function (t, e) { return t.__proto__ = e, t; }, _setPrototypeOf(t, e); }
import EventEmitter from 'eventemitter3';
import { debounce } from 'lodash';
import { RenderEvent, SceneEvent } from '../constant';
export var SceneRender = /*#__PURE__*/function (_EventEmitter) {
  function SceneRender(scene) {
    var _this;
    _classCallCheck(this, SceneRender);
    _this = _callSuper(this, SceneRender);
    _this.scene = void 0;
    _this.previousClick = void 0;
    /**
     * L7原生dblclick经常误触发，故改用这种方式
     * @param e
     */
    _this.onDblClick = function (e) {
      var _ref2, _e$pixel;
      var _ref = (_ref2 = (_e$pixel = e.pixel) !== null && _e$pixel !== void 0 ? _e$pixel : e.point) !== null && _ref2 !== void 0 ? _ref2 : {},
        _ref$x = _ref.x,
        x = _ref$x === void 0 ? 0 : _ref$x,
        _ref$y = _ref.y,
        y = _ref$y === void 0 ? 0 : _ref$y;
      var time = Date.now();
      if (_this.previousClick) {
        var _this$previousClick = _this.previousClick,
          oldX = _this$previousClick.x,
          oldY = _this$previousClick.y,
          oldTime = _this$previousClick.time;
        if (time - oldTime < 300 && Math.abs(x - oldX) < 5 && Math.abs(y - oldY) < 5) {
          _this.emit(RenderEvent.DblClick, e);
        }
      }
      _this.previousClick = {
        x: x,
        y: y,
        time: time
      };
    };
    _this.onMouseMove = function (e) {
      _this.emit(RenderEvent.Mousemove, e);
    };
    _this.onMouseDown = function (e) {
      _this.emit(RenderEvent.Dragstart, e);
    };
    _this.onDragging = function (e) {
      _this.emit(RenderEvent.Dragging, e);
    };
    _this.onDragEnd = debounce(function (e) {
      _this.emit(RenderEvent.Dragend, e);
    }, 0);
    _this.scene = scene;
    return _this;
  }
  _inherits(SceneRender, _EventEmitter);
  return _createClass(SceneRender, [{
    key: "enableDrag",
    value: function enableDrag() {
      this.disableDrag();
      this.scene.on(SceneEvent.Mousedown, this.onMouseDown);
      this.scene.on(SceneEvent.Dragging, this.onDragging);
      this.scene.on(SceneEvent.Mouseup, this.onDragEnd);
      this.scene.on(SceneEvent.Dragend, this.onDragEnd);
    }
  }, {
    key: "disableDrag",
    value: function disableDrag() {
      this.scene.off(SceneEvent.Mousedown, this.onMouseDown);
      this.scene.off(SceneEvent.Dragging, this.onDragging);
      this.scene.off(SceneEvent.Mouseup, this.onDragEnd);
      this.scene.off(SceneEvent.Dragend, this.onDragEnd);
    }
  }, {
    key: "enableMouseMove",
    value: function enableMouseMove() {
      this.disableMouseMove();
      this.scene.on(SceneEvent.Mousemove, this.onMouseMove);
    }
  }, {
    key: "disableMouseMove",
    value: function disableMouseMove() {
      this.scene.off(SceneEvent.Mousemove, this.onMouseMove);
    }
  }, {
    key: "enableDblClick",
    value: function enableDblClick() {
      this.disableDblClick();
      this.scene.on(SceneEvent.Mousedown, this.onDblClick);
    }
  }, {
    key: "disableDblClick",
    value: function disableDblClick() {
      this.scene.off(SceneEvent.Mousedown, this.onDblClick);
    }
  }]);
}(EventEmitter);
