// @antv/gpt-vis@0.4.7/dist/cjs/OrganizationChart/index.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/OrganizationChart/index.tsx
var OrganizationChart_exports = {};
__export(OrganizationChart_exports, {
  default: () => OrganizationChart_default
});
module.exports = __toCommonJS(OrganizationChart_exports);
var import_graphs = require("@ant-design/graphs");
var import_react = __toESM(require("react"));
var import_hooks = require("../ConfigProvider/hooks");
var import_graph = require("../utils/graph");
var { OrganizationChartNode } = import_graphs.RCNode;
var defaultConfig = {
  padding: [40, 0, 0, 120],
  autoFit: "view",
  autoResize: true,
  node: {
    style: {
      component: (d) => {
        var _a;
        const isActive = (_a = d.states) == null ? void 0 : _a.includes("active");
        return /* @__PURE__ */ import_react.default.createElement(
          OrganizationChartNode,
          {
            name: d.name,
            position: d.description,
            status: "online",
            isActive
          }
        );
      },
      size: [280, 80]
    }
  },
  edge: {
    state: {
      active: {
        stroke: "#1890ff",
        halo: false
      }
    }
  },
  behaviors: (prev) => [...prev, "hover-activate-neighbors"],
  transforms: (prev) => [
    ...prev.filter((t) => t.type !== "collapse-expand-react-node"),
    {
      ...prev.find(
        (t) => t.type === "collapse-expand-react-node"
      ),
      enable: true,
      iconOffsetY: 24
    }
  ],
  animation: false
};
var OrganizationChart = (props) => {
  const { data: propsData, ...restProps } = props;
  const data = (0, import_react.useMemo)(() => (0, import_graph.visTreeData2GraphData)(propsData), [propsData]);
  const config = (0, import_hooks.useGraphConfig)(
    "OrganizationChart",
    defaultConfig,
    restProps
  );
  return /* @__PURE__ */ import_react.default.createElement(import_graphs.OrganizationChart, { data, ...config });
};
var OrganizationChart_default = OrganizationChart;
