// @antv/gpt-vis@0.4.7/dist/cjs/export.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/export.ts
var export_exports = {};
__export(export_exports, {
  Area: () => import_Area.default,
  Bar: () => import_Bar.default,
  Column: () => import_Column.default,
  DEFAULT_CHART_COMPONENTS: () => DEFAULT_CHART_COMPONENTS,
  DualAxes: () => import_DualAxes.default,
  FishboneDiagram: () => import_FishboneDiagram.default,
  FlowDiagram: () => import_FlowDiagram.default,
  HeatMap: () => import_export_map2.HeatMap,
  Histogram: () => import_Histogram.default,
  IndentedTree: () => import_IndentedTree.default,
  Line: () => import_Line.default,
  Map: () => import_export_map.Map,
  MindMap: () => import_MindMap.default,
  NetworkGraph: () => import_NetworkGraph.default,
  OrganizationChart: () => import_OrganizationChart.default,
  PathMap: () => import_export_map2.PathMap,
  Pie: () => import_Pie.default,
  PinMap: () => import_export_map2.PinMap,
  Radar: () => import_Radar.default,
  Scatter: () => import_Scatter.default,
  Treemap: () => import_Treemap.default,
  VisText: () => import_Text.VisText,
  WordCloud: () => import_WordCloud.default
});
module.exports = __toCommonJS(export_exports);
var import_types = require("./types");
var import_Area = __toESM(require("./Area"));
var import_Bar = __toESM(require("./Bar"));
var import_Column = __toESM(require("./Column"));
var import_DualAxes = __toESM(require("./DualAxes"));
var import_Histogram = __toESM(require("./Histogram"));
var import_Line = __toESM(require("./Line"));
var import_Pie = __toESM(require("./Pie"));
var import_Radar = __toESM(require("./Radar"));
var import_Scatter = __toESM(require("./Scatter"));
var import_Treemap = __toESM(require("./Treemap"));
var import_WordCloud = __toESM(require("./WordCloud"));
var import_FishboneDiagram = __toESM(require("./FishboneDiagram"));
var import_FlowDiagram = __toESM(require("./FlowDiagram"));
var import_MindMap = __toESM(require("./MindMap"));
var import_NetworkGraph = __toESM(require("./NetworkGraph"));
var import_IndentedTree = __toESM(require("./IndentedTree"));
var import_OrganizationChart = __toESM(require("./OrganizationChart"));
var import_export_map = require("./export-map");
var import_export_map2 = require("./export-map");
var import_Text = require("./Text");
var DEFAULT_CHART_COMPONENTS = {
  [import_types.ChartType.Line]: import_Line.default,
  [import_types.ChartType.Column]: import_Column.default,
  [import_types.ChartType.Pie]: import_Pie.default,
  [import_types.ChartType.Bar]: import_Bar.default,
  [import_types.ChartType.Area]: import_Area.default,
  [import_types.ChartType.Scatter]: import_Scatter.default,
  [import_types.ChartType.PinMap]: import_export_map2.PinMap,
  [import_types.ChartType.HeatMap]: import_export_map2.HeatMap,
  [import_types.ChartType.MindMap]: import_MindMap.default,
  [import_types.ChartType.FlowDiagram]: import_FlowDiagram.default,
  [import_types.ChartType.NetworkGraph]: import_NetworkGraph.default
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Area,
  Bar,
  Column,
  DEFAULT_CHART_COMPONENTS,
  DualAxes,
  FishboneDiagram,
  FlowDiagram,
  HeatMap,
  Histogram,
  IndentedTree,
  Line,
  Map,
  MindMap,
  NetworkGraph,
  OrganizationChart,
  PathMap,
  Pie,
  PinMap,
  Radar,
  Scatter,
  Treemap,
  VisText,
  WordCloud
});
