// @antv/larkmap@1.5.1/lib/components/Legend/LegendCategories/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CLS_PREFIX = void 0;
exports.LegendCategories = LegendCategories;
var _react = _interopRequireDefault(require("react"));
var _classnames = _interopRequireDefault(require("classnames"));
var _color = require("../../../utils/color");
require("./index.css");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
const CLS_PREFIX = exports.CLS_PREFIX = 'larkmap-legend-category';
function LegendCategories(props) {
  const {
    labels,
    colors,
    geometryType = 'circle',
    isStrokeColor = false,
    style,
    className: cls_name
  } = props;
  function getColor(item) {
    return isStrokeColor ? {
      border: `2px solid ${item}`
    } : {
      background: item
    };
  }
  function Conent(color) {
    return /*#__PURE__*/_react.default.createElement("div", {
      className: (0, _classnames.default)(CLS_PREFIX, cls_name),
      style: style
    }, labels.map((item, index) => /*#__PURE__*/_react.default.createElement("div", {
      className: `${CLS_PREFIX}__content`,
      key: `${item}__${color[index]}__${index}`
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: (0, _classnames.default)(`${CLS_PREFIX}__content__shape`, {
        [`${CLS_PREFIX}__content__${geometryType}`]: geometryType
      }),
      style: getColor(color[index])
    }), /*#__PURE__*/_react.default.createElement("div", {
      className: `${CLS_PREFIX}labels`
    }, item))));
  }
  function Renders() {
    if (Array.isArray(colors)) {
      return Conent(colors);
    }
    const colorGradient = (0, _color.getGradientColors)(colors.startColor, colors.endColor, labels.length);
    return Conent(colorGradient);
  }
  return /*#__PURE__*/_react.default.createElement(Renders, null);
}
