// @antv/gpt-vis@0.4.7/dist/esm/Area/index.js
import { Area as ADCArea } from '@ant-design/plots';
import { get } from 'lodash';
import React from 'react';
import { usePlotConfig } from "../ConfigProvider/hooks";
var defaultConfig = function defaultConfig(props) {
  var data = props.data,
    _props$xField = props.xField,
    xField = _props$xField === void 0 ? 'time' : _props$xField,
    _props$yField = props.yField,
    yField = _props$yField === void 0 ? 'value' : _props$yField;
  var hasGroupField = get(data, '[0].group') !== undefined;
  var axisYTitle = get(props, 'axis.y.title');
  var defalutStyle = hasGroupField ? {} : {
    opacity: 0.6
  };
  return {
    xField: xField,
    yField: yField,
    style: defalutStyle,
    colorField: hasGroupField ? 'group' : undefined,
    tooltip: function tooltip(d) {
      var tooltipName = axisYTitle || d[xField];
      return {
        name: tooltipName,
        value: d[yField]
      };
    }
  };
};
var Area = function Area(props) {
  var config = usePlotConfig('Area', defaultConfig, props);
  return /*#__PURE__*/React.createElement(ADCArea, config);
};
export default Area;
