// @antv/gpt-vis@0.4.7/dist/esm/Text/mini-charts/proportion/index.js
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import { isNaN, isString, toNumber } from 'lodash';
import React, { useMemo } from 'react';
import { TEXT_THEME } from "../../theme";
import { useSvgWrapper } from "../hooks/useSvgWrapper";
import { getArcPath } from "./getArcPath";
function getProportionNumber(children,
// markdown 中传入的 children 文本为 string[]
origin) {
  var result;
  var originNumber = toNumber(origin);
  if (!isNaN(originNumber)) {
    result = originNumber;
  } else {
    var _text, _text$endsWith;
    var text;
    if (isString(children)) {
      text = children;
    } else if (Array.isArray(children) && isString(children[0])) {
      text = children[0];
    }
    if ((_text = text) !== null && _text !== void 0 && (_text$endsWith = _text.endsWith) !== null && _text$endsWith !== void 0 && _text$endsWith.call(_text, '%')) {
      var _text2;
      var percentageStr = (_text2 = text) === null || _text2 === void 0 ? void 0 : _text2.replace(/%$/, '');
      var proportionNumber = toNumber(percentageStr);
      if (!isNaN(proportionNumber)) result = proportionNumber / 100;
    }
  }
  return result;
}
export var ProportionChart = function ProportionChart(_ref) {
  var origin = _ref.origin,
    children = _ref.children;
  var data = useMemo(function () {
    return getProportionNumber(children, origin);
  }, [children, origin]);
  var _useSvgWrapper = useSvgWrapper(),
    _useSvgWrapper2 = _slicedToArray(_useSvgWrapper, 2),
    Svg = _useSvgWrapper2[0],
    fontSize = _useSvgWrapper2[1];
  var r = fontSize / 2;
  return data && /*#__PURE__*/React.createElement(Svg, {
    width: fontSize,
    height: fontSize
  }, /*#__PURE__*/React.createElement("circle", {
    cx: r,
    cy: r,
    r: r,
    fill: TEXT_THEME.chart.proportionShadowColor
  }), data >= 1 ? /*#__PURE__*/React.createElement("circle", {
    cx: r,
    cy: r,
    r: r,
    fill: TEXT_THEME.chart.proportionFillColor
  }) : /*#__PURE__*/React.createElement("path", {
    d: getArcPath(fontSize, data),
    fill: TEXT_THEME.chart.proportionFillColor
  }));
};
