// @antv/larkmap@1.5.1/lib/components/Control/MouseLocationControl/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MouseLocationControl = void 0;
var _l = require("@antv/l7");
var _lodash = require("lodash");
var _react = require("react");
var _utils = require("../../../utils");
var _hooks = require("../../LarkMap/hooks");
var _hooks2 = require("../hooks");
const MouseLocationControl = ({
  position,
  className,
  style,
  onLocationChange,
  transform,
  onAdd,
  onRemove,
  onShow,
  onHide
}) => {
  const scene = (0, _hooks.useScene)();
  const [control, setControl] = (0, _react.useState)();
  const styleText = (0, _react.useMemo)(() => (0, _utils.getStyleText)(style), [style]);
  const controlOptions = (0, _react.useMemo)(() => {
    return {
      transform,
      position,
      className,
      style: styleText
    };
  }, [transform, position, className, styleText]);
  (0, _react.useEffect)(() => {
    const mouseLocation = new _l.MouseLocation((0, _lodash.omitBy)(controlOptions, value => value === undefined));
    setControl(mouseLocation);
    scene.addControl(mouseLocation);
    return () => {
      scene.removeControl(mouseLocation);
      setControl(undefined);
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
  (0, _hooks2.useL7ComponentUpdate)(control, controlOptions);
  (0, _hooks2.useL7ComponentEvent)(control, {
    add: onAdd,
    remove: onRemove,
    show: onShow,
    hide: onHide,
    locationChange: onLocationChange
  });
  return null;
};
exports.MouseLocationControl = MouseLocationControl;
