// @antv/gpt-vis@0.4.7/dist/cjs/types/chart.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/types/chart.ts
var chart_exports = {};
__export(chart_exports, {
  ChartType: () => ChartType
});
module.exports = __toCommonJS(chart_exports);
var ChartType = /* @__PURE__ */ ((ChartType2) => {
  ChartType2["Pie"] = "pie";
  ChartType2["Column"] = "column";
  ChartType2["Line"] = "line";
  ChartType2["Area"] = "area";
  ChartType2["Scatter"] = "scatter";
  ChartType2["Histogram"] = "histogram";
  ChartType2["Treemap"] = "treemap";
  ChartType2["Bar"] = "bar";
  ChartType2["WordCloud"] = "word-cloud";
  ChartType2["DualAxes"] = "dual-axes";
  ChartType2["Radar"] = "radar";
  ChartType2["PinMap"] = "pin-map";
  ChartType2["PathMap"] = "path-map";
  ChartType2["HeatMap"] = "heat-map";
  ChartType2["MindMap"] = "mind-map";
  ChartType2["FishboneDiagram"] = "fishbone-diagram";
  ChartType2["FlowDiagram"] = "flow-diagram";
  ChartType2["IndentedTree"] = "indented-tree";
  ChartType2["NetworkGraph"] = "network-graph";
  ChartType2["OrganizationChart"] = "organization-chart";
  ChartType2["VisText"] = "vis-text";
  return ChartType2;
})(ChartType || {});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ChartType
});
