// @antv/gpt-vis@0.4.7/dist/esm/Pie/index.js
import { Pie as ADCPie } from '@ant-design/plots';
import { round, sumBy } from 'lodash';
import React from 'react';
import { usePlotConfig } from "../ConfigProvider/hooks";

/**
 * PieDataItem is the type for each data item in the pie chart.
 * It includes the name of the sector and its corresponding value.
 * @param category The name of the sector.
 * @param value The value of the sector.
 */

/**
 * the props for the Pie
 * @param data pie data
 */

var defaultConfig = function defaultConfig(props) {
  var _props$data = props.data,
    data = _props$data === void 0 ? [] : _props$data,
    _props$angleField = props.angleField,
    angleField = _props$angleField === void 0 ? 'value' : _props$angleField,
    _props$colorField = props.colorField,
    colorField = _props$colorField === void 0 ? 'category' : _props$colorField;
  var sumValue = sumBy(data, angleField);
  return {
    angleField: angleField,
    colorField: colorField,
    tooltip: function tooltip(d) {
      return {
        name: d[colorField],
        value: d[angleField]
      };
    },
    label: {
      position: 'outside',
      // text: angleField,
      text: function text(d) {
        return "".concat(d[colorField], ": ").concat(round(d[angleField] / sumValue * 100, 1), "%");
      }
    },
    legend: {
      color: {
        title: false,
        position: 'top'
      }
    },
    interaction: {
      elementSelect: {
        single: true
      }
    }
  };
};
var Pie = function Pie(props) {
  var config = usePlotConfig('Pie', defaultConfig, props);
  return /*#__PURE__*/React.createElement(ADCPie, config);
};
export default Pie;
