// @antv/gpt-vis@0.4.7/dist/cjs/utils/map/style.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/utils/map/style.ts
var style_exports = {};
__export(style_exports, {
  DefaultMarkerStyle: () => DefaultMarkerStyle,
  DefaultPolylineStyle: () => DefaultPolylineStyle,
  formatMakerStyle: () => formatMakerStyle,
  formatMapStyle: () => formatMapStyle,
  formatPolylineStyle: () => formatPolylineStyle
});
module.exports = __toCommonJS(style_exports);
var import_lodash = require("lodash");
var DefaultMapConfig = {
  mapOptions: {
    center: [120.210792, 30.246026],
    zoom: 16,
    maxZoom: 18,
    pitch: 0,
    rotation: 0,
    zoomEnable: true,
    pitchEnable: true
  }
};
var formatMapStyle = (props) => {
  const config = {
    mapOptions: {
      ...DefaultMapConfig.mapOptions,
      center: props.longitude && props.latitude ? [props.longitude, props.latitude] : DefaultMapConfig.mapOptions.center,
      zoom: props.scale || DefaultMapConfig.mapOptions.zoom,
      pitch: props.skew || DefaultMapConfig.mapOptions.pitch,
      rotation: props.rotate || DefaultMapConfig.mapOptions.rotation,
      zoomEnable: props.enableZoom || DefaultMapConfig.mapOptions.zoomEnable,
      pitchEnable: props.enableRotate || DefaultMapConfig.mapOptions.pitchEnable,
      // 地图底图样式
      style: props == null ? void 0 : props.mapType,
      // 高德地图密钥
      token: props == null ? void 0 : props.token
    },
    logoVisible: false
  };
  return config;
};
var DefaultMarkerStyle = {
  width: 12,
  anchorX: 0,
  anchorY: 1,
  label: {
    content: "",
    color: "#000000",
    fontSize: 10,
    borderRadius: 5,
    bgColor: "#ffffff",
    padding: 5
  },
  iconPath: "https://mdn.alipayobjects.com/huamei_qa8qxu/afts/img/A*3XdDTbsQ84QAAAAAAAAAAAAADmJ7AQ/original"
};
var formatMakerStyle = (data, markerStyle) => {
  const labelStyle = Object.assign({}, DefaultMarkerStyle.label, markerStyle == null ? void 0 : markerStyle.label);
  return data.map((marker, index) => {
    const { label, id, ...rest } = marker;
    return {
      ...DefaultMarkerStyle,
      ...markerStyle,
      ...rest,
      label: {
        ...labelStyle,
        ...(0, import_lodash.isObject)(label) ? label : { content: label }
      },
      id: id || index.toString()
    };
  });
};
var DefaultPolylineStyle = {
  width: 2,
  color: "#16f",
  dottedLine: false,
  zIndex: 1
};
var formatPolylineStyle = (data, polylineStyle) => {
  return data.map((item) => {
    return {
      ...DefaultPolylineStyle,
      ...polylineStyle,
      ...item
    };
  });
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  DefaultMarkerStyle,
  DefaultPolylineStyle,
  formatMakerStyle,
  formatMapStyle,
  formatPolylineStyle
});
