// @antv/larkmap@1.5.1/lib/components/Layers/CompositeLayers/FlowLayer/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FlowLayerEventMap = void 0;
var _constant = require("../../hooks/use-layer-event/constant");
const FlowLayerEventMap = exports.FlowLayerEventMap = {
  // 通用事件
  onRemove: _constant.LayerEventMap.onRemove,
  onShow: _constant.LayerEventMap.onShow,
  onHide: _constant.LayerEventMap.onHide,
  onDataUpdate: _constant.LayerEventMap.onDataUpdate,
  onLegend: _constant.LayerEventMap.onLegend,
  onLegendColor: _constant.LayerEventMap.onLegendColor,
  onLegendSize: _constant.LayerEventMap.onLegendSize,
  // 客流点点击事件
  onCircleLayerClick: 'circleLayer:' + _constant.LayerEventMap.onClick,
  onCircleLayerUnClick: 'circleLayer:' + _constant.LayerEventMap.onUnClick,
  onCircleLayerDblClick: 'circleLayer:' + _constant.LayerEventMap.onDblClick,
  onCircleLayerUndblclick: 'circleLayer:' + _constant.LayerEventMap.onUndblclick,
  onCircleLayerContextMenu: 'circleLayer:' + _constant.LayerEventMap.onContextMenu,
  onCircleLayerUnContextMenu: 'circleLayer:' + _constant.LayerEventMap.onUnContextMenu,
  onCircleLayerMouseEnter: 'circleLayer:' + _constant.LayerEventMap.onMouseEnter,
  onCircleLayerMouseMove: 'circleLayer:' + _constant.LayerEventMap.onMouseMove,
  onCircleLayerMouseOut: 'circleLayer:' + _constant.LayerEventMap.onMouseOut,
  onCircleLayerMouseUp: 'circleLayer:' + _constant.LayerEventMap.onMouseUp,
  onCircleLayerMouseDown: 'circleLayer:' + _constant.LayerEventMap.onMouseDown,
  onCircleLayerUnMousemove: 'circleLayer:' + _constant.LayerEventMap.onUnMousemove,
  onCircleLayerUnMouseup: 'circleLayer:' + _constant.LayerEventMap.onUnMouseup,
  onCircleLayerUnMousedown: 'circleLayer:' + _constant.LayerEventMap.onUnMousedown,
  onCircleLayerUnPick: 'circleLayer:' + _constant.LayerEventMap.onUnPick,
  // 客流线点击事件
  onLineLayerClick: 'lineLayer:' + _constant.LayerEventMap.onClick,
  onLineLayerUnClick: 'lineLayer:' + _constant.LayerEventMap.onUnClick,
  onLineLayerDblClick: 'lineLayer:' + _constant.LayerEventMap.onDblClick,
  onLineLayerUndblclick: 'lineLayer:' + _constant.LayerEventMap.onUndblclick,
  onLineLayerContextMenu: 'lineLayer:' + _constant.LayerEventMap.onContextMenu,
  onLineLayerUnContextMenu: 'lineLayer:' + _constant.LayerEventMap.onUnContextMenu,
  onLineLayerMouseEnter: 'lineLayer:' + _constant.LayerEventMap.onMouseEnter,
  onLineLayerMouseMove: 'lineLayer:' + _constant.LayerEventMap.onMouseMove,
  onLineLayerMouseOut: 'lineLayer:' + _constant.LayerEventMap.onMouseOut,
  onLineLayerMouseUp: 'lineLayer:' + _constant.LayerEventMap.onMouseUp,
  onLineLayerMouseDown: 'lineLayer:' + _constant.LayerEventMap.onMouseDown,
  onLineLayerUnMousemove: 'lineLayer:' + _constant.LayerEventMap.onUnMousemove,
  onLineLayerUnMouseup: 'lineLayer:' + _constant.LayerEventMap.onUnMouseup,
  onLineLayerUnMousedown: 'lineLayer:' + _constant.LayerEventMap.onUnMousedown,
  onLineLayerUnPick: 'lineLayer:' + _constant.LayerEventMap.onUnPick
};
