// @antv/gpt-vis@0.4.7/dist/esm/utils/graph.js
import _objectSpread from "@babel/runtime/helpers/esm/objectSpread2";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
var _excluded = ["children"];
// FIXME: 全量导入 G6 模块
import { G6 } from '@ant-design/graphs';
var treeToGraphData = G6.treeToGraphData;
export function visTreeData2GraphData(data) {
  return treeToGraphData(data, {
    getNodeData: function getNodeData(datum, depth) {
      datum.id = datum.name;
      datum.depth = depth;
      if (!datum.children) return datum;
      var children = datum.children,
        restDatum = _objectWithoutProperties(datum, _excluded);
      return _objectSpread(_objectSpread({}, restDatum), {}, {
        children: children.map(function (child) {
          return child.name;
        })
      });
    },
    getEdgeData: function getEdgeData(source, target) {
      return {
        source: source.name,
        target: target.name
      };
    }
  });
}
export function visGraphData2GraphData(data) {
  return {
    nodes: data.nodes.map(function (node) {
      return {
        id: node.name,
        style: {
          labelText: node.name
        }
      };
    }),
    edges: data.edges.map(function (edge) {
      return {
        source: edge.source,
        target: edge.target,
        style: {
          labelText: edge.name
        }
      };
    })
  };
}
