// @antv/larkmap@1.5.1/lib/components/Control/LogoControl/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LogoControl = void 0;
var _l = require("@antv/l7");
var _lodash = require("lodash");
var _react = require("react");
var _utils = require("../../../utils");
var _hooks = require("../../LarkMap/hooks");
var _hooks2 = require("../hooks");
const LogoControl = ({
  img,
  href,
  position,
  className,
  style,
  onShow,
  onHide,
  onAdd,
  onRemove
}) => {
  const scene = (0, _hooks.useScene)();
  const [control, setControl] = (0, _react.useState)();
  const styleText = (0, _react.useMemo)(() => (0, _utils.getStyleText)(style), [style]);
  const controlOptions = (0, _react.useMemo)(() => {
    return {
      img,
      href,
      position,
      className,
      style: styleText
    };
  }, [img, href, position, className, styleText]);
  (0, _react.useEffect)(() => {
    const logo = new _l.Logo((0, _lodash.omitBy)(controlOptions, value => value === undefined));
    setControl(logo);
    scene.addControl(logo);
    return () => {
      scene.removeControl(logo);
      setControl(undefined);
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
  (0, _hooks2.useL7ComponentUpdate)(control, controlOptions);
  (0, _hooks2.useL7ComponentEvent)(control, {
    add: onAdd,
    remove: onRemove,
    show: onShow,
    hide: onHide
  });
  return null;
};
exports.LogoControl = LogoControl;
