// @antv/gpt-vis@0.4.7/dist/cjs/FlowDiagram/index.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/FlowDiagram/index.tsx
var FlowDiagram_exports = {};
__export(FlowDiagram_exports, {
  default: () => FlowDiagram_default
});
module.exports = __toCommonJS(FlowDiagram_exports);
var import_graphs = require("@ant-design/graphs");
var import_react = __toESM(require("react"));
var import_hooks = require("../ConfigProvider/hooks");
var import_config = require("../utils/config");
var import_graph = require("../utils/graph");
var import_helper = require("./helper");
var { TextNode } = import_graphs.RCNode;
var defaultConfig = {
  autoResize: true,
  autoFit: "view",
  node: {
    style: {
      component: (d) => {
        var _a;
        const isActive = (_a = d.states) == null ? void 0 : _a.includes("active");
        return /* @__PURE__ */ import_react.default.createElement(
          TextNode,
          {
            text: d.id,
            isActive,
            type: "filled",
            style: {
              fontSize: 12
            },
            borderWidth: 2
          }
        );
      },
      size: [140, 32]
    }
  },
  edge: {
    style: {
      endArrow: true,
      labelBackground: true,
      labelMaxLines: 2,
      labelMaxWidth: "40%",
      labelWordWrap: true
    },
    state: {
      active: {
        halo: false,
        labelWordWrap: false,
        stroke: "#001f98"
      }
    }
  },
  behaviors: (prev) => [
    ...prev,
    {
      type: "hover-activate-neighbors",
      onHover: (e) => {
        e.view.setCursor("pointer");
      },
      onHoverEnd: (e) => {
        e.view.setCursor("default");
      }
    }
  ]
};
var FlowDiagram = (props) => {
  const { data: propsData, ...restProps } = props;
  const data = (0, import_react.useMemo)(() => (0, import_graph.visGraphData2GraphData)(propsData), [propsData]);
  const config = (0, import_hooks.useGraphConfig)(
    "FlowDiagram",
    (0, import_config.mergeGraphOptions)(defaultConfig, (0, import_helper.getGraphOptionsByData)(data)),
    restProps
  );
  return /* @__PURE__ */ import_react.default.createElement(import_graphs.FlowGraph, { data, ...config });
};
var FlowDiagram_default = FlowDiagram;
