// @antv/gpt-vis@0.4.7/dist/esm/DualAxes/index.js
import _objectSpread from "@babel/runtime/helpers/esm/objectSpread2";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
var _excluded = ["categories", "series"];
import { DualAxes as ADCDualAxes } from '@ant-design/plots';
import React, { useMemo } from 'react';
import { usePlotConfig } from "../ConfigProvider/hooks";
import { transform } from "./util";
var defaultConfig = function defaultConfig(props) {
  var _props$xField = props.xField,
    xField = _props$xField === void 0 ? 'category' : _props$xField,
    axisXTitle = props.axisXTitle,
    legendTypeList = props.legendTypeList;
  return {
    xField: xField,
    axis: {
      x: {
        title: axisXTitle
      }
    },
    legend: {
      color: {
        itemMarker: function itemMarker(v, i) {
          return legendTypeList[i];
        }
      }
    },
    scale: {
      color: {
        palette: 'category10'
      }
    }
  };
};
var DualAxes = function DualAxes(props) {
  var categories = props.categories,
    series = props.series,
    others = _objectWithoutProperties(props, _excluded);
  var transformData = useMemo(function () {
    return transform(series, categories);
  }, [props]);
  var config = usePlotConfig('DualAxes', defaultConfig, _objectSpread(_objectSpread({}, others), transformData));
  return /*#__PURE__*/React.createElement(ADCDualAxes, config);
};
export default DualAxes;
