// @antv/l7-composite-layers@0.17.1/dist/lib/composite-layers/flow-layer/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_OPTIONS = exports.DEFAULT_FLOW_LAYER_SELECT_OPTIONS = exports.DEFAULT_FLOW_LAYER_ACTIVE_OPTIONS = exports.EMPTY_LINE_LAYER_SOURCE = exports.EMPTY_CIRCLE_LAYER_SOURCE = void 0;
exports.EMPTY_CIRCLE_LAYER_SOURCE = {
    data: [],
    parser: {
        type: 'json',
        x: 'lng',
        y: 'lat',
    },
};
exports.EMPTY_LINE_LAYER_SOURCE = {
    data: [],
    parser: {
        type: 'json',
        x: 'fromLng',
        y: 'fromLat',
        x1: 'toLng',
        y1: 'toLat',
    },
};
exports.DEFAULT_FLOW_LAYER_ACTIVE_OPTIONS = {
    enableCircleSpread: false,
    enableLineSpread: false,
    circleStrokeColor: '#ccc',
    lineStroke: '#ccc',
};
exports.DEFAULT_FLOW_LAYER_SELECT_OPTIONS = Object.assign(Object.assign({}, exports.DEFAULT_FLOW_LAYER_ACTIVE_OPTIONS), { enableCircleSpread: true, enableLineSpread: true });
exports.DEFAULT_OPTIONS = {
    source: {
        data: [],
        parser: {
            type: 'json',
            x: 'f_lon',
            y: 'f_lat',
            x1: 't_lon',
            y1: 't_lat',
            weight: 'weight',
        },
    },
    enableCluster: true,
    clusterType: 'HCA',
    clusterZoomStep: 1,
    clusterNodeSize: 64,
    clusterRadius: 40,
    clusterExtent: 512,
    maxTopFlowNum: 5000,
    circleColor: '#fff',
    circleRadius: {
        field: 'weight',
        value: [1, 16],
    },
    circleStrokeColor: '#000',
    circleStrokeWidth: 1,
    lineColor: {
        field: 'weight',
        value: ['#2a5674', '#d1eeea'],
    },
    lineWidth: {
        field: 'weight',
        value: [1, 16],
    },
    lineStroke: '#000',
    lineStrokeWidth: 1,
    fadeOpacityEnabled: true,
    fadeOpacityAmount: 0,
    showLocationName: false,
    locationNameSize: 12,
    locationNameColor: '#fff',
    locationNameStroke: '#000',
    locationNameStrokeWidth: 1,
    state: {
        active: exports.DEFAULT_FLOW_LAYER_ACTIVE_OPTIONS,
        select: exports.DEFAULT_FLOW_LAYER_SELECT_OPTIONS,
    },
};
