// @antv/larkmap@1.5.1/lib/components/ContextMenu/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ContextMenu = void 0;
var _classnames = _interopRequireDefault(require("classnames"));
var _react = _interopRequireWildcard(require("react"));
var _useScene = require("../LarkMap/hooks/use-scene");
var _Marker = require("../Marker");
var _constant = require("./constant");
var _ContextMenuItem = require("./ContextMenuItem");
require("./index.css");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
const ContextMenu = props => {
  const scene = (0, _useScene.useScene)();
  const [initMenu, setInitMenu] = (0, _react.useState)({
    visible: false,
    position: undefined
  });
  const initMenuRender = (0, _react.useMemo)(() => {
    return /*#__PURE__*/_react.default.createElement("div", {
      className: (0, _classnames.default)(`${_constant.CLS_PREFIX}`, props?.className),
      style: props?.style
    }, props?.children);
  }, [props]);

  // 右键打开面板信息
  const mapRightMenuOpen = e => {
    const {
      lng,
      lat
    } = e.lnglat;
    setInitMenu({
      visible: true,
      position: {
        lng,
        lat
      }
    });
  };

  // 单击事件关闭菜单
  const mapRightMenuClose = () => {
    const timeOut = setTimeout(() => {
      if (timeOut) {
        clearTimeout(timeOut);
        setInitMenu({
          visible: false,
          position: undefined
        });
      }
    }, 0);
  };
  (0, _react.useEffect)(() => {
    if (scene) {
      scene.on('contextmenu', mapRightMenuOpen);
      scene.on('click', mapRightMenuClose);
    }
    return () => {
      scene.off('contextmenu', mapRightMenuOpen);
      scene.off('click', mapRightMenuClose);
    };
  }, [scene]);
  return initMenu.visible ? /*#__PURE__*/_react.default.createElement(_Marker.Marker, {
    lngLat: initMenu.position,
    anchor: "top-left"
  }, initMenuRender) : null;
};
exports.ContextMenu = ContextMenu;
ContextMenu.Item = _ContextMenuItem.ContextMenuItem;
