// @antv/larkmap@1.5.1/lib/components/LocationSearch/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LocationSearch = void 0;
var _ahooks = require("ahooks");
var _md = _interopRequireDefault(require("md5"));
var _react = _interopRequireWildcard(require("react"));
var _utils = require("../../utils");
var _constant = require("./constant");
require("./index.css");
var _Select = _interopRequireDefault(require("./Select"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
const {
  Option
} = _Select.default;
const defaultSelectProps = {
  placeholder: '请输入要搜索地名',
  showSearch: true,
  allowClear: true,
  filterOption: false,
  defaultActiveFirstOption: false
};
const LocationSearch = ({
  searchParams,
  showDistrict = true,
  showAddress = true,
  onSearchFinish,
  onChange,
  ...selectProps
}) => {
  const [options, setOptions] = (0, _react.useState)([]);
  (0, _react.useEffect)(() => {
    onSearchFinish?.(options);
  }, [onSearchFinish, options]);
  const {
    run: onSearch
  } = (0, _ahooks.useDebounceFn)(async searchText => {
    if (!searchText) {
      setOptions([]);
      return;
    }
    const {
      keywords,
      privateKey,
      ...otherParams
    } = searchParams;
    const params = {
      ...otherParams,
      keywords: [...(keywords ?? '').split('|'), searchText].filter(item => !!item).join('|')
    };
    if (privateKey) {
      params.sig = (0, _md.default)((0, _utils.getQueryString)(params, false) + privateKey);
    }
    const url = `${_constant.GAO_DE_API_URL}?${(0, _utils.getQueryString)(params, true)}`;
    const res = await (await fetch(url)).json();
    setOptions((res?.tips ?? []).filter(item => item.location && item.location.length).map(item => {
      const [lon, lat] = item.location.split(',');
      item.longitude = +lon;
      item.latitude = +lat;
      return item;
    }));
  }, {
    wait: 1000
  });
  const onLocationChange = (0, _react.useCallback)(name => {
    const targetOption = name && options.find(option => option.name === name);
    onChange?.(name || undefined, targetOption || undefined);
  }, [onChange, options]);
  return /*#__PURE__*/_react.default.createElement(_Select.default, _extends({
    className: `${_constant.CLS_PREFIX}`,
    onSearch: onSearch,
    onChange: onLocationChange,
    clearIcon: () => null
  }, defaultSelectProps, selectProps), options.map(option => {
    const tip = `${showDistrict ? option.district : ''}${showAddress ? option.address : ''}`;
    return /*#__PURE__*/_react.default.createElement(Option, {
      key: option.id,
      value: option.name
    }, /*#__PURE__*/_react.default.createElement("div", {
      title: option.name,
      className: `${_constant.CLS_PREFIX}__option-name`
    }, option.name), tip && /*#__PURE__*/_react.default.createElement("div", {
      title: tip,
      className: `${_constant.CLS_PREFIX}__option-tip`
    }, tip));
  }));
};
exports.LocationSearch = LocationSearch;
