// @antv/l7-draw@3.1.5/lib/mode/mid-point-mode.js
"use strict";

function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MidPointMode = void 0;
var _turf = require("@turf/turf");
var _constant = require("../constant");
var _utils = require("../utils");
var _pointMode = require("./point-mode");
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _classCallCheck(a, n) { if (!(a instanceof n)) throw new TypeError("Cannot call a class as a function"); }
function _defineProperties(e, r) { for (var t = 0; t < r.length; t++) { var o = r[t]; o.enumerable = o.enumerable || !1, o.configurable = !0, "value" in o && (o.writable = !0), Object.defineProperty(e, _toPropertyKey(o.key), o); } }
function _createClass(e, r, t) { return r && _defineProperties(e.prototype, r), t && _defineProperties(e, t), Object.defineProperty(e, "prototype", { writable: !1 }), e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
function _callSuper(t, o, e) { return o = _getPrototypeOf(o), _possibleConstructorReturn(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], _getPrototypeOf(t).constructor) : o.apply(t, e)); }
function _possibleConstructorReturn(t, e) { if (e && ("object" == _typeof(e) || "function" == typeof e)) return e; if (void 0 !== e) throw new TypeError("Derived constructors may only return object or undefined"); return _assertThisInitialized(t); }
function _assertThisInitialized(e) { if (void 0 === e) throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); return e; }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
function _get() { return _get = "undefined" != typeof Reflect && Reflect.get ? Reflect.get.bind() : function (e, t, r) { var p = _superPropBase(e, t); if (p) { var n = Object.getOwnPropertyDescriptor(p, t); return n.get ? n.get.call(arguments.length < 3 ? e : r) : n.value; } }, _get.apply(null, arguments); }
function _superPropBase(t, o) { for (; !{}.hasOwnProperty.call(t, o) && null !== (t = _getPrototypeOf(t));); return t; }
function _getPrototypeOf(t) { return _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf.bind() : function (t) { return t.__proto__ || Object.getPrototypeOf(t); }, _getPrototypeOf(t); }
function _inherits(t, e) { if ("function" != typeof e && null !== e) throw new TypeError("Super expression must either be null or a function"); t.prototype = Object.create(e && e.prototype, { constructor: { value: t, writable: !0, configurable: !0 } }), Object.defineProperty(t, "prototype", { writable: !1 }), e && _setPrototypeOf(t, e); }
function _setPrototypeOf(t, e) { return _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function (t, e) { return t.__proto__ = e, t; }, _setPrototypeOf(t, e); }
var MidPointMode = exports.MidPointMode = /*#__PURE__*/function (_PointMode) {
  function MidPointMode() {
    _classCallCheck(this, MidPointMode);
    return _callSuper(this, MidPointMode, arguments);
  }
  _inherits(MidPointMode, _PointMode);
  return _createClass(MidPointMode, [{
    key: "midPointRender",
    get:
    /**
     * 获取midPoint类型对应的render
     * @protected
     */
    function get() {
      return this.render.midPoint;
    }
  }, {
    key: "getCommonOptions",
    value: function getCommonOptions(options) {
      // @ts-ignore
      return _objectSpread(_objectSpread({}, _get(_getPrototypeOf(MidPointMode.prototype), "getCommonOptions", this).call(this, options)), {}, {
        showMidPoint: true
      });
    }
  }, {
    key: "bindMidPointRenderEvent",
    value: function bindMidPointRenderEvent() {
      var _this$midPointRender, _this$midPointRender2, _this$midPointRender3;
      (_this$midPointRender = this.midPointRender) === null || _this$midPointRender === void 0 ? void 0 : _this$midPointRender.on(_constant.RenderEvent.Click, this.onMidPointClick.bind(this));
      (_this$midPointRender2 = this.midPointRender) === null || _this$midPointRender2 === void 0 ? void 0 : _this$midPointRender2.on(_constant.RenderEvent.Mousemove, this.onMidPointHover.bind(this));
      (_this$midPointRender3 = this.midPointRender) === null || _this$midPointRender3 === void 0 ? void 0 : _this$midPointRender3.on(_constant.RenderEvent.Mouseout, this.onMidPointUnHover.bind(this));
    }
    /**
     * 获取中点数据
     */
  }, {
    key: "getMidPointData",
    value: function getMidPointData() {
      return this.source.getRenderData('midPoint');
    }
    /**
     * 设置中点数据
     * @param data
     */
  }, {
    key: "setMidPointData",
    value: function setMidPointData(data) {
      return this.source.setRenderData('midPoint', data);
    }
    /**
     * 计算并返回传入线段的中点数组
     * @param line
     */
  }, {
    key: "getMidPointsByLine",
    value: function getMidPointsByLine(line) {
      var nodes = line.properties.nodes;
      if (!this.options.showMidPoint || nodes.length < 2) {
        return [];
      }
      var midPoints = [];
      for (var index = 0; index < nodes.length - 1; index++) {
        var _nodes$index$properti, _nodes$index$properti2, _nodes$properties$id, _nodes$properties;
        var newMidPoint = (0, _turf.center)((0, _turf.featureCollection)([nodes[index], nodes[index + 1]]), {
          properties: {
            id: (0, _utils.getUuid)('midPoint'),
            startId: (_nodes$index$properti = (_nodes$index$properti2 = nodes[index].properties) === null || _nodes$index$properti2 === void 0 ? void 0 : _nodes$index$properti2.id) !== null && _nodes$index$properti !== void 0 ? _nodes$index$properti : '',
            endId: (_nodes$properties$id = (_nodes$properties = nodes[index + 1].properties) === null || _nodes$properties === void 0 ? void 0 : _nodes$properties.id) !== null && _nodes$properties$id !== void 0 ? _nodes$properties$id : ''
          }
        });
        midPoints.push(newMidPoint);
      }
      return midPoints;
    }
  }, {
    key: "onMidPointHover",
    value: function onMidPointHover(e) {
      this.setCursor('pointHover');
    }
  }, {
    key: "onMidPointUnHover",
    value: function onMidPointUnHover(e) {
      this.resetCursor();
    }
  }, {
    key: "enableMidPointRenderAction",
    value: function enableMidPointRenderAction() {
      var _this$midPointRender4, _this$midPointRender5;
      (_this$midPointRender4 = this.midPointRender) === null || _this$midPointRender4 === void 0 ? void 0 : _this$midPointRender4.enableClick();
      (_this$midPointRender5 = this.midPointRender) === null || _this$midPointRender5 === void 0 ? void 0 : _this$midPointRender5.enableHover();
    }
  }, {
    key: "disableMidPointRenderAction",
    value: function disableMidPointRenderAction() {
      var _this$midPointRender6, _this$midPointRender7;
      (_this$midPointRender6 = this.midPointRender) === null || _this$midPointRender6 === void 0 ? void 0 : _this$midPointRender6.disableClick();
      (_this$midPointRender7 = this.midPointRender) === null || _this$midPointRender7 === void 0 ? void 0 : _this$midPointRender7.disableHover();
    }
  }]);
}(_pointMode.PointMode);
