// @antv/l7-draw@3.1.5/lib/utils/common.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.transLngLat2Position = exports.splitByPosition = exports.isDev = exports.getPosition = exports.getMapDom = exports.getLngLat = exports.findMinIndex = void 0;
var _lodash = require("lodash");
// @ts-ignore
var isDev = exports.isDev = process.env.NODE_ENV === 'development';
/**
 * 获取完全覆盖地图区域的DOM，会根据地图类型返回不同的结果
 * @param scene
 */
var getMapDom = exports.getMapDom = function getMapDom(scene) {
  var _ref, _ref2, _ref3, _ref4, _ref5, _ref6, _container$querySelec;
  var container = scene.getContainer();
  return (_ref = (_ref2 = (_ref3 = (_ref4 = (_ref5 = (_ref6 = (_container$querySelec = container === null || container === void 0 ? void 0 : container.querySelector('.l7-marker-container')) !== null && _container$querySelec !== void 0 ? _container$querySelec : container === null || container === void 0 ? void 0 : container.querySelector('.BMap_mask')) !== null && _ref6 !== void 0 ? _ref6 : container === null || container === void 0 ? void 0 : container.querySelector('#tdt-L7')) !== null && _ref5 !== void 0 ? _ref5 : scene.getMapCanvasContainer()) !== null && _ref4 !== void 0 ? _ref4 : container === null || container === void 0 ? void 0 : container.querySelector('.l7-scene')) !== null && _ref3 !== void 0 ? _ref3 : container === null || container === void 0 ? void 0 : container.querySelector('.l7-control-container')) !== null && _ref2 !== void 0 ? _ref2 : container === null || container === void 0 ? void 0 : container.querySelector('.l7-marker-container2')) !== null && _ref !== void 0 ? _ref : null;
};
/**
 * 磨平L7 Scene 鼠标事件返回的经纬度差异
 * @param e
 */
var getLngLat = exports.getLngLat = function getLngLat(e) {
  // @ts-ignore
  return e.lngLat || e.lnglat || e.latlng;
};
var getPosition = exports.getPosition = function getPosition(e) {
  var _getLngLat = getLngLat(e),
    lng = _getLngLat.lng,
    lat = _getLngLat.lat;
  return [lng, lat];
};
/**
 * 将lnglat转换为position格式
 * @param lng
 * @param lat
 */
var transLngLat2Position = exports.transLngLat2Position = function transLngLat2Position(_ref7) {
  var lng = _ref7.lng,
    lat = _ref7.lat;
  return [lng, lat];
};
/**
 * 找到最小值的下标
 * @param array
 */
var findMinIndex = exports.findMinIndex = function findMinIndex(array) {
  var maxValue = Number.MAX_SAFE_INTEGER;
  var maxIndex = 0;
  var length = array.length;
  for (var index = 0; index < length; index++) {
    if (array[index] < maxValue) {
      maxValue = array[index];
      maxIndex = index;
    }
  }
  return maxIndex;
};
var splitByPosition = exports.splitByPosition = function splitByPosition(positions, splitPosition) {
  var linePositionsList = [];
  var linePositions = [];
  positions.forEach(function (position) {
    if (!(0, _lodash.isEqual)(position, splitPosition)) {
      linePositions.push(position);
    } else if (linePositions.length) {
      linePositionsList.push(linePositions);
      linePositions = [];
    }
  });
  if (linePositions.length) {
    linePositionsList.push(linePositions);
  }
  return linePositionsList;
};
