// @antv/l7-draw@3.1.5/es/drawer/line-drawer.js
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
function _slicedToArray(r, e) { return _arrayWithHoles(r) || _iterableToArrayLimit(r, e) || _unsupportedIterableToArray(r, e) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(r, a) { if (r) { if ("string" == typeof r) return _arrayLikeToArray(r, a); var t = {}.toString.call(r).slice(8, -1); return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? Array.from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0; } }
function _arrayLikeToArray(r, a) { (null == a || a > r.length) && (a = r.length); for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e]; return n; }
function _iterableToArrayLimit(r, l) { var t = null == r ? null : "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (null != t) { var e, n, i, u, a = [], f = !0, o = !1; try { if (i = (t = t.call(r)).next, 0 === l) { if (Object(t) !== t) return; f = !1; } else for (; !(f = (e = i.call(t)).done) && (a.push(e.value), a.length !== l); f = !0); } catch (r) { o = !0, n = r; } finally { try { if (!f && null != t.return && (u = t.return(), Object(u) !== u)) return; } finally { if (o) throw n; } } return a; } }
function _arrayWithHoles(r) { if (Array.isArray(r)) return r; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _classCallCheck(a, n) { if (!(a instanceof n)) throw new TypeError("Cannot call a class as a function"); }
function _defineProperties(e, r) { for (var t = 0; t < r.length; t++) { var o = r[t]; o.enumerable = o.enumerable || !1, o.configurable = !0, "value" in o && (o.writable = !0), Object.defineProperty(e, _toPropertyKey(o.key), o); } }
function _createClass(e, r, t) { return r && _defineProperties(e.prototype, r), t && _defineProperties(e, t), Object.defineProperty(e, "prototype", { writable: !1 }), e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
function _callSuper(t, o, e) { return o = _getPrototypeOf(o), _possibleConstructorReturn(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], _getPrototypeOf(t).constructor) : o.apply(t, e)); }
function _possibleConstructorReturn(t, e) { if (e && ("object" == _typeof(e) || "function" == typeof e)) return e; if (void 0 !== e) throw new TypeError("Derived constructors may only return object or undefined"); return _assertThisInitialized(t); }
function _assertThisInitialized(e) { if (void 0 === e) throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); return e; }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
function _get() { return _get = "undefined" != typeof Reflect && Reflect.get ? Reflect.get.bind() : function (e, t, r) { var p = _superPropBase(e, t); if (p) { var n = Object.getOwnPropertyDescriptor(p, t); return n.get ? n.get.call(arguments.length < 3 ? e : r) : n.value; } }, _get.apply(null, arguments); }
function _superPropBase(t, o) { for (; !{}.hasOwnProperty.call(t, o) && null !== (t = _getPrototypeOf(t));); return t; }
function _getPrototypeOf(t) { return _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf.bind() : function (t) { return t.__proto__ || Object.getPrototypeOf(t); }, _getPrototypeOf(t); }
function _inherits(t, e) { if ("function" != typeof e && null !== e) throw new TypeError("Super expression must either be null or a function"); t.prototype = Object.create(e && e.prototype, { constructor: { value: t, writable: !0, configurable: !0 } }), Object.defineProperty(t, "prototype", { writable: !1 }), e && _setPrototypeOf(t, e); }
function _setPrototypeOf(t, e) { return _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function (t, e) { return t.__proto__ = e, t; }, _setPrototypeOf(t, e); }
import { coordAll } from '@turf/turf';
import { last } from 'lodash';
import { DrawEvent, RenderEvent } from '../constant';
import { LineMode } from '../mode';
import { createDashLine, createPointFeature, getDefaultLineProperties, getPosition, isSameFeature, joinMultiFeatures, splitMultiFeatures } from '../utils';
export var LineDrawer = /*#__PURE__*/function (_LineMode) {
  function LineDrawer(scene, options) {
    var _this;
    _classCallCheck(this, LineDrawer);
    _this = _callSuper(this, LineDrawer, [scene, options]);
    _this.drawLineFinish = function () {
      var _drawLine$properties$;
      var drawLine = _this.drawLine;
      var nodes = (_drawLine$properties$ = drawLine === null || drawLine === void 0 ? void 0 : drawLine.properties.nodes) !== null && _drawLine$properties$ !== void 0 ? _drawLine$properties$ : [];
      if (!drawLine || (nodes === null || nodes === void 0 ? void 0 : nodes.length) <= 1) {
        return;
      }
      _this.setActiveLine(drawLine);
      var _this$options = _this.options,
        autoActive = _this$options.autoActive,
        editable = _this$options.editable;
      if (!autoActive || !editable) {
        _this.handleLineUnClick(drawLine);
      }
      if (editable && autoActive) {
        _this.setHelper('pointHover');
      } else {
        _this.setHelper(_this.addable ? 'draw' : null);
      }
      _this.emit(DrawEvent.Add, drawLine, _this.getLineData());
    };
    _this.sceneRender.on(RenderEvent.DblClick, _this.drawLineFinish);
    _this.bindPointRenderEvent();
    _this.bindSceneEvent();
    _this.bindMidPointRenderEvent();
    _this.bindLineRenderEvent();
    return _this;
  }
  _inherits(LineDrawer, _LineMode);
  return _createClass(LineDrawer, [{
    key: "dragItem",
    get: function get() {
      return this.dragLine;
    }
  }, {
    key: "editItem",
    get: function get() {
      return this.editLine;
    }
  }, {
    key: "getMainLayer",
    value: function getMainLayer() {
      var _this$render$line$get, _this$render$line;
      return (_this$render$line$get = (_this$render$line = this.render.line) === null || _this$render$line === void 0 ? void 0 : _this$render$line.getLayers()) !== null && _this$render$line$get !== void 0 ? _this$render$line$get : [];
    }
  }, {
    key: "setData",
    value: function setData(lines) {
      var lineFeatures = splitMultiFeatures(lines).map(function (line) {
        var _line$properties, _line$properties$node;
        line.properties = _objectSpread(_objectSpread({}, getDefaultLineProperties()), (_line$properties = line.properties) !== null && _line$properties !== void 0 ? _line$properties : {});
        if (!((_line$properties$node = line.properties.nodes) === null || _line$properties$node === void 0 ? void 0 : _line$properties$node.length)) {
          line.properties.nodes = coordAll(line).map(function (position) {
            return createPointFeature(position);
          });
        }
        return line;
      });
      this.source.setData({
        point: [],
        midPoint: [],
        dashLine: [],
        line: lineFeatures
      });
      this.setTextData(this.getAllTexts());
      if (this.editLine) {
        this.setActiveLine(this.editLine);
      }
    }
  }, {
    key: "getData",
    value: function getData() {
      var getOriginData = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : false;
      var lines = this.getLineData();
      if (getOriginData) {
        return lines;
      }
      return joinMultiFeatures(lines);
    }
  }, {
    key: "getRenderTypes",
    value: function getRenderTypes() {
      return ['line', 'dashLine', 'midPoint', 'point', 'text'];
    }
  }, {
    key: "bindPointRenderEvent",
    value: function bindPointRenderEvent() {
      var _this$pointRender;
      _get(_getPrototypeOf(LineDrawer.prototype), "bindPointRenderEvent", this).call(this);
      (_this$pointRender = this.pointRender) === null || _this$pointRender === void 0 ? void 0 : _this$pointRender.on(RenderEvent.Contextmenu, this.onPointContextMenu.bind(this));
    }
  }, {
    key: "onPointClick",
    value: function onPointClick(e) {
      var _drawLine$properties$2,
        _this2 = this;
      var drawLine = this.drawLine;
      var nodes = (_drawLine$properties$2 = drawLine === null || drawLine === void 0 ? void 0 : drawLine.properties.nodes) !== null && _drawLine$properties$2 !== void 0 ? _drawLine$properties$2 : [];
      var feature = e.feature;
      if (isSameFeature(feature, last(nodes))) {
        requestAnimationFrame(function () {
          _this2.drawLineFinish();
        });
      } else {
        var _feature$geometry$coo = _slicedToArray(feature.geometry.coordinates, 2),
          lng = _feature$geometry$coo[0],
          lat = _feature$geometry$coo[1];
        e.lngLat = {
          lng: lng,
          lat: lat
        };
        this.onPointCreate(e);
      }
    }
  }, {
    key: "removeNode",
    value: function removeNode(node, feature) {
      var _targetFeature$proper;
      var targetFeature = this.getTargetFeature(feature);
      var targetNode = this.getTargetFeature(node, (_targetFeature$proper = targetFeature === null || targetFeature === void 0 ? void 0 : targetFeature.properties.nodes) !== null && _targetFeature$proper !== void 0 ? _targetFeature$proper : []);
      if (targetFeature && targetNode) {
        var _targetFeature$proper2;
        var nodes = (_targetFeature$proper2 = targetFeature === null || targetFeature === void 0 ? void 0 : targetFeature.properties.nodes) !== null && _targetFeature$proper2 !== void 0 ? _targetFeature$proper2 : [];
        if (nodes.length < 3) {
          return;
        }
        this.syncLineNodes(targetFeature, nodes.filter(function (node) {
          return !isSameFeature(targetNode, node);
        }));
        this.emit(DrawEvent.RemoveNode, targetNode, targetFeature, this.getLineData());
        this.emit(DrawEvent.Edit, targetFeature, this.getLineData());
      }
    }
  }, {
    key: "onPointContextMenu",
    value: function onPointContextMenu(e) {
      var _editLine$properties$;
      var editLine = this.editLine;
      var deleteNode = e.feature;
      var nodes = (_editLine$properties$ = editLine === null || editLine === void 0 ? void 0 : editLine.properties.nodes) !== null && _editLine$properties$ !== void 0 ? _editLine$properties$ : [];
      if (!editLine || nodes.length < 3) {
        return;
      }
      this.removeNode(deleteNode, editLine);
      this.scene.setMapStatus({
        dragEnable: true
      });
      return deleteNode;
    }
  }, {
    key: "onPointCreate",
    value: function onPointCreate(e) {
      if (!this.addable) {
        return;
      }
      var feature = _get(_getPrototypeOf(LineDrawer.prototype), "onPointCreate", this).call(this, e);
      if (feature) {
        this.setHelper('drawFinish');
        this.emit(DrawEvent.AddNode, feature, this.drawLine, this.getLineData());
      }
      return feature;
    }
  }, {
    key: "onPointDragEnd",
    value: function onPointDragEnd(e) {
      var editLine = this.editLine;
      var feature = _get(_getPrototypeOf(LineDrawer.prototype), "onPointDragEnd", this).call(this, e);
      if (editLine && feature) {
        this.emit(DrawEvent.Edit, editLine, this.getLineData());
      }
      return feature;
    }
  }, {
    key: "onLineDragStart",
    value: function onLineDragStart(e) {
      var feature = _get(_getPrototypeOf(LineDrawer.prototype), "onLineDragStart", this).call(this, e);
      if (feature) {
        this.emit(DrawEvent.DragStart, feature, this.getLineData());
      }
      return feature;
    }
  }, {
    key: "onLineDragging",
    value: function onLineDragging(e) {
      var feature = _get(_getPrototypeOf(LineDrawer.prototype), "onLineDragging", this).call(this, e);
      if (feature) {
        this.emit(DrawEvent.Dragging, feature, this.getLineData());
      }
      return feature;
    }
  }, {
    key: "onLineDragEnd",
    value: function onLineDragEnd(e) {
      var feature = _get(_getPrototypeOf(LineDrawer.prototype), "onLineDragEnd", this).call(this, e);
      if (feature) {
        this.emit(DrawEvent.DragEnd, feature, this.getLineData());
        this.emit(DrawEvent.Edit, feature, this.getLineData());
      }
      return feature;
    }
  }, {
    key: "onMidPointClick",
    value: function onMidPointClick(e) {
      var editLine = this.editLine;
      var feature = _get(_getPrototypeOf(LineDrawer.prototype), "onMidPointClick", this).call(this, e);
      if (editLine && feature) {
        this.emit(DrawEvent.Edit, editLine, this.getLineData());
        this.emit(DrawEvent.AddNode, feature, editLine, this.getLineData());
      }
      return feature;
    }
  }, {
    key: "onSceneMouseMove",
    value: function onSceneMouseMove(e) {
      var drawLine = this.drawLine;
      if (!drawLine) {
        return;
      }
      var lastNode = last(drawLine.properties.nodes);
      var mousePosition = getPosition(e);
      if (this.options.adsorbOptions) {
        var _this$getAdsorbPositi;
        mousePosition = (_this$getAdsorbPositi = this.getAdsorbPosition(mousePosition)) !== null && _this$getAdsorbPositi !== void 0 ? _this$getAdsorbPositi : mousePosition;
      }
      this.setDashLineData([createDashLine([mousePosition, lastNode.geometry.coordinates])]);
      this.setTextData(this.getAllTexts());
    }
  }, {
    key: "onSceneDragEnd",
    value: function onSceneDragEnd(e) {
      if (!this.isDragTrigger || !this.addable || !this.drawLine) {
        return;
      }
      this.drawLineFinish();
    }
  }, {
    key: "setActiveFeature",
    value: function setActiveFeature(target) {
      var targetFeature = this.getTargetFeature(target);
      if (targetFeature) {
        this.setActiveLine(targetFeature);
      } else {
        this.clearActiveLine();
      }
    }
  }, {
    key: "resetFeatures",
    value: function resetFeatures() {
      var features = this.getLineData();
      if (this.drawLine) {
        features = features.filter(function (feature) {
          return !feature.properties.isDraw;
        });
        this.source.setData({
          point: [],
          dashLine: [],
          midPoint: []
        });
        this.setTextData(function (features) {
          return features.filter(function (feature) {
            return !feature.properties.isActive;
          });
        });
      }
      if (this.editLine) {
        this.handleLineUnClick(this.editLine);
      }
      this.setLineData(features.map(function (feature) {
        feature.properties = _objectSpread(_objectSpread({}, feature.properties), {}, {
          isDrag: false,
          isActive: false,
          isHover: false
        });
        return feature;
      }));
    }
  }, {
    key: "enablePointRenderAction",
    value: function enablePointRenderAction() {
      _get(_getPrototypeOf(LineDrawer.prototype), "enablePointRenderAction", this).call(this);
      if (this.options.editable) {
        var _this$pointRender2;
        (_this$pointRender2 = this.pointRender) === null || _this$pointRender2 === void 0 ? void 0 : _this$pointRender2.enableContextMenu();
      }
    }
  }, {
    key: "disablePointRenderAction",
    value: function disablePointRenderAction() {
      var _this$pointRender3;
      _get(_getPrototypeOf(LineDrawer.prototype), "disablePointRenderAction", this).call(this);
      (_this$pointRender3 = this.pointRender) === null || _this$pointRender3 === void 0 ? void 0 : _this$pointRender3.disableContextMenu();
    }
  }, {
    key: "bindEnableEvent",
    value: function bindEnableEvent() {
      _get(_getPrototypeOf(LineDrawer.prototype), "bindEnableEvent", this).call(this);
      this.enableSceneRenderAction();
      this.enablePointRenderAction();
      this.enableLineRenderAction();
      this.enableMidPointRenderAction();
      this.sceneRender.enableDblClick();
    }
  }, {
    key: "unbindEnableEvent",
    value: function unbindEnableEvent() {
      _get(_getPrototypeOf(LineDrawer.prototype), "unbindEnableEvent", this).call(this);
      this.disableSceneRenderAction();
      this.disablePointRenderAction();
      this.disableLineRenderAction();
      this.disableMidPointRenderAction();
      this.sceneRender.disableDblClick();
    }
  }, {
    key: "bindThis",
    value: function bindThis() {
      _get(_getPrototypeOf(LineDrawer.prototype), "bindThis", this).call(this);
      this.bindPointRenderEvent = this.bindPointRenderEvent.bind(this);
      this.bindSceneEvent = this.bindSceneEvent.bind(this);
      this.bindLineRenderEvent = this.bindLineRenderEvent.bind(this);
      this.bindMidPointRenderEvent = this.bindMidPointRenderEvent.bind(this);
    }
  }]);
}(LineMode);
