// @antv/l7-draw@3.1.5/es/interactive/popup.js
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _classCallCheck(a, n) { if (!(a instanceof n)) throw new TypeError("Cannot call a class as a function"); }
function _defineProperties(e, r) { for (var t = 0; t < r.length; t++) { var o = r[t]; o.enumerable = o.enumerable || !1, o.configurable = !0, "value" in o && (o.writable = !0), Object.defineProperty(e, _toPropertyKey(o.key), o); } }
function _createClass(e, r, t) { return r && _defineProperties(e.prototype, r), t && _defineProperties(e, t), Object.defineProperty(e, "prototype", { writable: !1 }), e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
import Tippy from 'tippy.js';
import { getMapDom } from '../utils';
import 'tippy.js/dist/tippy.css';
import 'tippy.js/themes/light.css';
import { DEFAULT_POPUP_CONFIG, SceneEvent } from '../constant';
import { debounce } from 'lodash';
export var Popup = /*#__PURE__*/function () {
  function Popup(scene, tippyProps) {
    var _this = this;
    _classCallCheck(this, Popup);
    this.content = '';
    this.tippy = void 0;
    this.scene = void 0;
    this.isMouseInner = false;
    this.onMouseMove = function () {
      _this.isMouseInner = true;
      _this.checkTippyShow();
    };
    this.onMouseOut = function () {
      _this.isMouseInner = false;
      _this.checkTippyShow();
    };
    this.setContent = debounce(function (content) {
      var _this$tippy;
      _this.content = content !== null && content !== void 0 ? content : '';
      (_this$tippy = _this.tippy) === null || _this$tippy === void 0 ? void 0 : _this$tippy.setContent(content !== null && content !== void 0 ? content : '');
      _this.checkTippyShow();
    }, 16, {
      maxWait: 16
    });
    var container = getMapDom(scene);
    this.scene = scene;
    if (container) {
      this.tippy = Tippy(getMapDom(scene), _objectSpread(_objectSpread({}, DEFAULT_POPUP_CONFIG), tippyProps));
      this.tippy.hide();
      scene.on(SceneEvent.Mousemove, this.onMouseMove);
      scene.on(SceneEvent.Mouseout, this.onMouseOut);
    }
  }
  return _createClass(Popup, [{
    key: "getContent",
    value: function getContent() {
      return this.content;
    }
  }, {
    key: "checkTippyShow",
    value: function checkTippyShow() {
      if (this.content && this.isMouseInner) {
        var _this$tippy2;
        (_this$tippy2 = this.tippy) === null || _this$tippy2 === void 0 ? void 0 : _this$tippy2.show();
      } else {
        var _this$tippy3;
        (_this$tippy3 = this.tippy) === null || _this$tippy3 === void 0 ? void 0 : _this$tippy3.hide();
      }
    }
  }, {
    key: "destroy",
    value: function destroy() {
      var _this$tippy4;
      this.scene.off(SceneEvent.Mousemove, this.onMouseMove);
      this.scene.off(SceneEvent.Mouseout, this.onMouseOut);
      (_this$tippy4 = this.tippy) === null || _this$tippy4 === void 0 ? void 0 : _this$tippy4.destroy();
    }
  }]);
}();
