// @antv/l7-composite-layers@0.17.1/dist/lib/core/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LayerGroupEvent = exports.CompositeLayerEvent = exports.CoreLayerEvent = exports.OriginLayerEventList = exports.OriginMouseLayerEventList = void 0;
exports.OriginMouseLayerEventList = [
    // 鼠标事件
    // 单击事件
    'click',
    'dblclick',
    'unclick',
    'undblclick',
    // 右键事件
    'contextmenu',
    'uncontextmenu',
    // mouse 事件
    'mouseenter',
    'unmouseenter',
    'mousemove',
    'unmousemove',
    'mouseout',
    'mouseup',
    'unmouseup',
    'mousedown',
    'unpick',
];
exports.OriginLayerEventList = [
    // 图层事件
    // 代理掉的事件 'inited', 'add', 'remove',
    'show',
    'hide',
    'dataUpdate',
    'destroy',
    // 数据映射更新，图例更新事件
    'legend',
    'legend:color',
    'legend:size',
    ...exports.OriginMouseLayerEventList,
];
var CoreLayerEvent;
(function (CoreLayerEvent) {
    CoreLayerEvent["CREATED"] = "created";
    CoreLayerEvent["INITED"] = "inited";
    CoreLayerEvent["ADD"] = "add";
    CoreLayerEvent["REMOVE"] = "remove";
})(CoreLayerEvent || (exports.CoreLayerEvent = CoreLayerEvent = {}));
var CompositeLayerEvent;
(function (CompositeLayerEvent) {
    CompositeLayerEvent["CREATED"] = "created";
    CompositeLayerEvent["INITED"] = "inited";
    CompositeLayerEvent["ADD"] = "add";
    CompositeLayerEvent["REMOVE"] = "remove";
})(CompositeLayerEvent || (exports.CompositeLayerEvent = CompositeLayerEvent = {}));
var LayerGroupEvent;
(function (LayerGroupEvent) {
    LayerGroupEvent["INITED_LAYER"] = "inited-layer";
    LayerGroupEvent["INITED_LAYERS"] = "inited-layers";
})(LayerGroupEvent || (exports.LayerGroupEvent = LayerGroupEvent = {}));
