// @antv/gpt-vis@0.4.7/dist/esm/Bar/index.js
import { Bar as ADCBar } from '@ant-design/plots';
import { get } from 'lodash';
import React from 'react';
import { usePlotConfig } from "../ConfigProvider/hooks";
var defaultConfig = function defaultConfig(props) {
  var data = props.data,
    _props$xField = props.xField,
    xField = _props$xField === void 0 ? 'category' : _props$xField,
    _props$yField = props.yField,
    yField = _props$yField === void 0 ? 'value' : _props$yField;
  var hasGroupField = get(data, '[0].group') !== undefined;
  var axisYTitle = get(props, 'axis.y.title');
  return {
    xField: xField,
    yField: yField,
    colorField: hasGroupField ? 'group' : undefined,
    tooltip: function tooltip(d) {
      var tooltipName = axisYTitle || d[xField];
      return {
        name: tooltipName,
        value: d[yField]
      };
    },
    style: {
      // 圆角样式
      radiusTopLeft: 5,
      radiusTopRight: 5
    }
  };
};
var Bar = function Bar(props) {
  var config = usePlotConfig('Bar', defaultConfig, props);
  return /*#__PURE__*/React.createElement(ADCBar, config);
};
export default Bar;
