// @antv/larkmap@1.5.1/es/components/Control/LayerSwitchControl/index.js
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : String(i); }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }
function _iterableToArrayLimit(r, l) { var t = null == r ? null : "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (null != t) { var e, n, i, u, a = [], f = !0, o = !1; try { if (i = (t = t.call(r)).next, 0 === l) { if (Object(t) !== t) return; f = !1; } else for (; !(f = (e = i.call(t)).done) && (a.push(e.value), a.length !== l); f = !0); } catch (r) { o = !0, n = r; } finally { try { if (!f && null != t.return && (u = t.return(), Object(u) !== u)) return; } finally { if (o) throw n; } } return a; } }
function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }
import { LayerSwitch as L7LayerSwitch } from '@antv/l7';
import { omitBy } from 'lodash-es';
import React, { useEffect, useMemo, useState } from 'react';
import { getStyleText } from "../../../utils";
import { useLayerList, useScene } from "../../LarkMap/hooks";
import { useL7ComponentEvent, useL7ComponentPortal, useL7ComponentUpdate } from "../hooks";
export var LayerSwitchControl = function LayerSwitchControl(_ref) {
  var layerItems = _ref.layers,
    popperPlacement = _ref.popperPlacement,
    popperTrigger = _ref.popperTrigger,
    popperClassName = _ref.popperClassName,
    btnIcon = _ref.btnIcon,
    btnText = _ref.btnText,
    title = _ref.title,
    vertical = _ref.vertical,
    position = _ref.position,
    className = _ref.className,
    style = _ref.style,
    multiple = _ref.multiple,
    onAdd = _ref.onAdd,
    onRemove = _ref.onRemove,
    onShow = _ref.onShow,
    onHide = _ref.onHide,
    onPopperShow = _ref.onPopperShow,
    onPopperHide = _ref.onPopperHide,
    onSelectChange = _ref.onSelectChange;
  var scene = useScene();
  var fullLayerList = useLayerList();
  var _useState = useState(),
    _useState2 = _slicedToArray(_useState, 2),
    control = _useState2[0],
    setControl = _useState2[1];
  var styleText = useMemo(function () {
    return getStyleText(style);
  }, [style]);
  var _useL7ComponentPortal = useL7ComponentPortal(btnIcon),
    btnIconPortal = _useL7ComponentPortal.portal,
    btnIconDOM = _useL7ComponentPortal.dom;
  var layers = useMemo(function () {
    return layerItems !== null && layerItems !== void 0 && layerItems.length ? layerItems.map(function (layerItem) {
      if (typeof layerItem === 'string') {
        return fullLayerList.find(function (layer) {
          return layer.id === layerItem;
        });
      }
      // 当 layerItem 为 LayerSwitchItem 类型
      if (!Object.prototype.hasOwnProperty.call(layerItem, 'isComposite')) {
        var _ref2 = layerItem,
          layer = _ref2.layer;
        if (typeof layer === 'string') {
          var targetLayer = fullLayerList.find(function (item) {
            return item.id === layer;
          });
          return targetLayer ? _objectSpread(_objectSpread({}, layerItem), {}, {
            layer: targetLayer
          }) : undefined;
        }
        return layerItem;
      }
      return layerItem;
    }).filter(function (layer) {
      return !!layer;
    }) : fullLayerList;
  }, [layerItems, fullLayerList]);
  var controlOptions = useMemo(function () {
    return {
      layers: layers,
      popperPlacement: popperPlacement,
      popperTrigger: popperTrigger,
      popperClassName: popperClassName,
      btnText: btnText,
      title: title,
      vertical: vertical,
      position: position,
      className: className,
      multiple: multiple,
      style: styleText,
      btnIcon: btnIconDOM
    };
  }, [layers, popperPlacement, popperTrigger, popperClassName, btnText, title, vertical, position, className, multiple, styleText, btnIconDOM]);
  useEffect(function () {
    var layerSwitch = new L7LayerSwitch(omitBy(controlOptions, function (value) {
      return value === undefined;
    }));
    setControl(layerSwitch);
    scene.addControl(layerSwitch);
    return function () {
      scene.removeControl(layerSwitch);
      setControl(undefined);
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
  useL7ComponentUpdate(control, controlOptions);
  useL7ComponentEvent(control, {
    add: onAdd,
    remove: onRemove,
    show: onShow,
    hide: onHide,
    popperShow: onPopperShow,
    popperHide: onPopperHide,
    selectChange: onSelectChange
  });
  return /*#__PURE__*/React.createElement(React.Fragment, null, btnIconPortal);
};
